ingest/cas.py

```diff
@@ -3,24 +3,204 @@
+
+# Comment out block below below unless debugging
+# from cellarium.cas.visualization import CASCircularTreePlotUMAPDashApp
+
+with suppress_stderr():
+    cl = CellOntologyCache()
+
+input_path = "pbmc_10x_v3_4k.h5ad"
+cas_response_output_path = f"{input_path}__cas_response_ontology_aware.json"
+
+def cas_annotate(input_path, output_path):
+    """Call CAS for an H5AD, write results; return input AnnData, CAS response
+    """
+    # TODO (SCP-5715):
+    # - Add CAS API token to Google Secrets Manager
+    # - Update block below to use GSM
+    # - Ensure team removes any local .cas-api-token files and .gitignore entry
+    with open(".cas-api-token") as f:
+        api_token = f.read().strip()
+
+    cas = CASClient(api_token=api_token)
+    print('CASClient')
+    print(cas)
+
+    adata = sc.read_h5ad(input_path)
+
+    # Returns summary results that have substantial post-processing support
+    cas_ontology_aware_response = cas.annotate_matrix_cell_type_ontology_aware_strategy(
+        matrix=adata,
+        chunk_size=500,
+        feature_ids_column_name='gene_ids',
+        feature_names_column_name='index'
+    )
+
+    # Returns dataset-specific results.
+    # Ontology-aware strategy (above, not this) is preferred.
+    # cas_response = cas.annotate_matrix_cell_type_summary_statistics_strategy(
+    #     matrix=adata
+    # )
+
+    with open(output_path, "w") as f:
+        f.write(json.dumps(cas_ontology_aware_response))
+
+    adata.write(f"cas_annotated_before_postprocessing__{input_path}")
+
+    print(f"Wrote CAS response to: {output_path}")
+
+    return [adata, cas_ontology_aware_response]
+
+def make_compliant_for_scp(adata):
+    """Shim to make AnnData file compliant with SCP metadata schema
+
+    Only use for demo / development purposes.  This is generally commented out
+    upstream.
+    """
+    if 'biosample_id' not in adata.obs:
+        adata.obs['biosample_id'] = "sample-1"
+        adata.obs['donor_id'] = "donor-1"
+        adata.obs['species'] = "NCBITaxon_9606"
+        adata.obs['species__ontology_label'] = "Homo sapiens"
+        adata.obs['disease'] = "PATO_0000461"
+        adata.obs['disease__ontology_label'] = "normal"
+        adata.obs['organ'] = "UBERON_0000178"
+        adata.obs['organ__ontology_label'] = "blood"
+        adata.obs['library_preparation_protocol'] = "EFO_0030059"
+        adata.obs['library_preparation_protocol__ontology_label'] = "10x 3' v3"
+        adata.obs['sex'] = "female"
+    return adata
+
+# Stub for potential later expansion
+# def format_as_scp_metadatum(cas_response_path):
+#     with open(cas_response_filepath) as f:
+#         cas_response = json.loads(f.read())
+
+#     tsv_rows = []
+#     for cas_item in cas_response:
+#         cell = cas_item["query_cell_id"]
+#         first_match = cas_item["matches"][0]
+#         annotation_label = cas_item["matches"][0]
+
+def merge_cas_results(adata, cas_ontology_aware_response):
```

Comment:
For the study owner-requested CAS use case, merging definitely makes sense. Exploring extracting CAS results as a clustering file is another avenue, especially if we find a UX that involves enabling user-requested CAS.

Reply:
Agreed, this seems worth more exploration in SCP-5753 and / or [CAS-SCP engineering strategy](https://docs.google.com/document/d/11eNlwMjwKoUcx4TQflzxw4DuN9q03fNNLphUuAVNIc4/edit#heading=h.elzs70nakpsf).